File: QadeerApp.Web/Modules/ServerTypes/Administration/UserForm.ts
import { BooleanEditor, EmailAddressEditor, ImageUploadEditor, initFormType, LookupEditor, PasswordEditor, PrefixedContext, StringEditor } from "@serenity-is/corelib";

export interface UserForm {
    Username: StringEditor;
    DisplayName: StringEditor;
    DepartmentId: LookupEditor;
    SpecializationId: LookupEditor;
    Email: EmailAddressEditor;
    Roles: LookupEditor;
    UserImage: ImageUploadEditor;
    Password: PasswordEditor;
    PasswordConfirm: PasswordEditor;
    Source: StringEditor;
    IsActive: BooleanEditor;
}

export class UserForm extends PrefixedContext {
    static readonly formKey = 'Administration.User';
    private static init: boolean;

    constructor(prefix: string) {
        super(prefix);

        if (!UserForm.init) {
            UserForm.init = true;

            var w0 = StringEditor;
            var w1 = LookupEditor;
            var w2 = EmailAddressEditor;
            var w3 = ImageUploadEditor;
            var w4 = PasswordEditor;
            var w5 = BooleanEditor;

            initFormType(UserForm, [
                'Username', w0,
                'DisplayName', w0,
                'DepartmentId', w1,
                'SpecializationId', w1,
                'Email', w2,
                'Roles', w1,
                'UserImage', w3,
                'Password', w4,
                'PasswordConfirm', w4,
                'Source', w0,
                'IsActive', w5
            ]);
        }
    }
}

File: QadeerApp.Web/Modules/Administration/User/UserForm.cs
namespace QadeerApp.Administration.Forms;

[FormScript("Administration.User")]
[BasedOnRow(typeof(UserRow), CheckNames = true)]
public class UserForm
{
    [LabelWidth(200, UntilNext = true)]
    public string Username { get; set; }
    public string DisplayName { get; set; }
    [LookupEditor(typeof(DepartmentRow), InplaceAdd = false)]
    [Required(true)]
    public int DepartmentId { get; set; }
    [LookupEditor(typeof(SpecializationRow), CascadeFrom = "DepartmentId", CascadeField = "DepartmentId", InplaceAdd = false)]
    public int? SpecializationId { get; set; }
    [EmailAddressEditor]
    public string Email { get; set; }
    [LookupEditor(typeof(RoleRow), Multiple = true)]
    public List<int> Roles { get; set; }
    public string UserImage { get; set; }
    [PasswordEditor, Required(true)]
    public string Password { get; set; }
    [PasswordEditor, Required(true)]
    public string PasswordConfirm { get; set; }
    [SkipOnSave]
    public string Source { get; set; }
    public bool IsActive { get; set; }
}
